src/daemon/rpmostreed-transaction-livefs.c

```diff
@@ -0,0 +1,688 @@
+    rpmostree_syscore_get_origin_merge_deployment (sysroot,
+                                                   ostree_deployment_get_osname (booted_deployment));
+  g_assert (origin_merge_deployment);
+  const char *booted_csum = ostree_deployment_get_csum (booted_deployment);
+  const char *target_csum = ostree_deployment_get_csum (origin_merge_deployment);
+
+  /* Require a pending deployment to use as a source - perhaps in the future we
+   * handle direct live overlays.
+   */
+  if (origin_merge_deployment == booted_deployment)
+    return glnx_throw (error, "No pending deployment");
+
+  /* Open a fd for the booted deployment */
+  g_autofree char *deployment_path = ostree_sysroot_get_deployment_dirpath (sysroot, booted_deployment);
+  glnx_fd_close int deployment_dfd = -1;
+  if (!glnx_opendirat (ostree_sysroot_get_fd (sysroot), deployment_path, TRUE,
+                       &deployment_dfd, error))
+    return FALSE;
+
+  /* Find out whether we already have a live overlay */
+  g_autofree char *live_inprogress = NULL;
+  g_autofree char *live_replaced = NULL;
+  if (!rpmostree_syscore_deployment_get_live (sysroot, booted_deployment, -1,
+                                              &live_inprogress, &live_replaced,
+                                              error))
+    return FALSE;
+  const char *resuming_overlay = NULL;
+  if (live_inprogress != NULL)
+    {
+      if (strcmp (live_inprogress, target_csum) == 0)
+        resuming_overlay = target_csum;
+    }
+  const char *replacing_overlay = NULL;
+  if (live_replaced != NULL)
+    {
+      if (strcmp (live_replaced, target_csum) == 0)
+        return glnx_throw (error, "Current overlay is already %s", target_csum);
+      replacing_overlay = live_replaced;
+    }
+
+  if (resuming_overlay)
+    g_print ("Note: Resuming interrupted overlay of %s\n", target_csum);
+  if (replacing_overlay)
+    g_print ("Note: Previous overlay: %s\n", replacing_overlay);
+
+  /* Look at the difference between the two commits - we could also walk the
+   * filesystem, but doing it at the ostree level is potentially faster, since
+   * we know when two directories are the same.
+   */
+  g_autoptr(CommitDiff) diff = NULL;
+  if (!analyze_commit_diff (repo, booted_csum, target_csum,
+                            &diff, cancellable, error))
+    return FALSE;
+
+  print_commit_diff (diff);
+
+  const gboolean replacing = (self->flags & RPMOSTREE_TRANSACTION_LIVEFS_FLAG_REPLACE) > 0;
+  const gboolean requires_etc_merge = (diff->flags & COMMIT_DIFF_FLAGS_ETC) > 0;
+  const gboolean adds_packages = diff->added_pkgs->len > 0;
+  const gboolean modifies_packages = diff->removed_pkgs->len > 0 || diff->modified_pkgs_new->len > 0;
+  if (!adds_packages)
+    return glnx_throw (error, "No packages added; live updates not currently supported for modifications or deletions");
+  /* Is this a dry run? */
+  if ((self->flags & RPMOSTREE_TRANSACTION_LIVEFS_FLAG_DRY_RUN) > 0)
```

Comment:
I'd bump this lower down, after the other error checks. So that if the dry run passes, then the real thing should pass as well.